lib/python/picongpu/extra/input/preparingInsightData.py

```diff
@@ -0,0 +1,472 @@
+        for i in range(self.w.size):
+            interp_FF = RegularGridInterpolator((a * scale[i], b * scale[i]), FF[:, :, i], method=method)
+            self.Ew[:, :, i] = interp_FF((Y, X))
+
+        self.Ew = self.Ew / np.abs(self.Ew).max()  # rescale amplitude to 1
+
+    def correctPhase(self, GVD=0, TOD=0):
+        """
+        Since there is no information about the global phase of every wavelength measured with Insight,
+        there has to be a phase correction.
+        For that, we assume perfect compression of the beam center in the near field, thus we substract
+        a global phase for every wavelength.
+        Furthermore, there is the possibility to add dispersion parameters (GVD, TOD) to the data.
+
+        Arguments:
+        GVD : Group velocity disperion in fs**2/rad,  optional. The default is 0.
+        TOD : Third order dispersion in fs**3/rad**2, optional. The default is 0.
+        """
+        # phase in near field beam center
+        phase_centr = np.unwrap(
+            np.angle(self.Ew_NF[np.abs(self.y_NF - self.yc_NF).argmin(), np.abs(self.x_NF - self.xc_NF).argmin(), :])
+        )
+        dispersion = -GVD / 2 * (self.w - self.w[self.wc_idx]) ** 2 - TOD / 6 * (self.w - self.w[self.wc_idx]) ** 3
+
+        self.Ew_NF = np.abs(self.Ew_NF) * np.exp(
+            1j * (np.angle(self.Ew_NF) - phase_centr + dispersion)
+        )  # correcting near field
+        self.Ew = np.abs(self.Ew) * np.exp(1j * (np.angle(self.Ew) - phase_centr + dispersion))  # correcting far field
+
+        self.isPhaseCorrected = True
+        print(
+            "Phase has been corrected. \nApplied dispersion parameters: GVD = %.f fs**2/rad, TOD = %.f fs**3/rad**2"
+            % (GVD, TOD)
+        )
+
+    def correctUglySpotInNF(self, x_ugly, y_ugly, uglybins=3, method="linear"):
+        """
+        In case there is an ugly spot in the near field (such as an unnatural peak), it can be smoothened out with this function.
+        It automatically also corrects the far field by propagating the corrected near field back into the focus.
+
+        Arguments:
+        x_ugly: (near field) x coordinate of the ugly spot
+        y_ugly: (near field) y coordinate of the ugly spot
+        uglybins: number of bins surrounding the ugly spot which will be cut away / smoothened out (default is 3)
+        method: RegularGridInterpolator interpolation method,
+                default is 'linear' (since 'cubic' can cause some bumps in amplitude or phase)
+        """
+        # index borders of hole
+        x_ugly_idx = np.abs(self.x_NF - x_ugly).argmin()
+        y_ugly_idx = np.abs(self.y_NF - y_ugly).argmin()
+        xidx_low = x_ugly_idx - uglybins
+        xidx_upp = x_ugly_idx + uglybins
+        yidx_low = y_ugly_idx - uglybins
+        yidx_upp = y_ugly_idx + uglybins
+
+        x_surr = np.concatenate((self.x_NF[xidx_low - 3 : xidx_low], self.x_NF[x_ugly_idx + uglybins : xidx_upp + 3]))
```

Comment:
This magic `3` is another safety margin? Would you add a comment explaining that number please? Also, the rule of thumb 'Whatever you use twice, automate!' applies here too...

Reply:
And why do you apply it only along the x direction? (Comment in code please)